services/src/main/java/org/keycloak/authentication/DefaultAuthenticationFlow.java

```diff
@@ -373,7 +374,20 @@ private boolean conditionalNotMatched(AuthenticationExecutionModel model, List<A
         ConditionalAuthenticator authenticator = (ConditionalAuthenticator) createAuthenticator(factory);
         AuthenticationProcessor.Result context = processor.createAuthenticatorContext(model, authenticator, executionList);
 
-        return !authenticator.matchCondition(context);
+       boolean matchCondition;
+
+        // Retrieve previous evaluation result if any, else evaluate and store result for future re-evaluation
+        if (processor.isEvaluatedTrue(model)) {
+            matchCondition = true;
+        } else if (processor.isEvaluatedFalse(model)) {
+            matchCondition = false;
+        } else {
+            matchCondition = authenticator.matchCondition(context);
+            processor.getAuthenticationSession().setExecutionStatus(model.getId(),
+                    matchCondition?AuthenticationSessionModel.ExecutionStatus.EVALUATED_TRUE:AuthenticationSessionModel.ExecutionStatus.EVALUATED_FALSE);
```

Comment:
Minor: Is it possible to use spaces around "?" and ":" characters to improve readability a bit?

Reply:
No problem, I will do it.

Reply:
Done